File: backend/app/schemas/instructor.py
"""
Instructor profile schemas for InstaInstru platform.

This module defines Pydantic schemas for instructor-related operations,
including profile management and service offerings.

Note: The old booking-related fields (buffer_time, minimum_advance_hours,
default_session_duration) have been removed as part of the refactoring.
These will be reimplemented differently in the new booking system.
"""

from datetime import datetime
import logging
from typing import Any, List, Optional, Sequence

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_serializer

from ..core.constants import (
    MAX_BIO_LENGTH,
    MAX_SESSION_DURATION,
    MIN_BIO_LENGTH,
    MIN_SESSION_DURATION,
)
from ._strict_base import StrictRequestModel
from .address import ServiceAreaNeighborhoodOut
from .base import Money, StandardizedModel

logger = logging.getLogger(__name__)


class InstructorFilterParams(BaseModel):
    """
    Query parameters for filtering instructors.

    All fields are optional to maintain backward compatibility.
    Supports text search, skill filtering, and price range filtering.
    """

    search: Optional[str] = Field(
        None,
        description="Text search across instructor name, bio, and services",
        min_length=1,
        max_length=100,
    )
    service_catalog_id: Optional[str] = Field(
        None, description="Filter by specific service catalog ID"
    )
    min_price: Optional[float] = Field(
        None, ge=0, le=1000, description="Minimum hourly rate filter"
    )
    max_price: Optional[float] = Field(
        None, ge=0, le=1000, description="Maximum hourly rate filter"
    )
    age_group: Optional[str] = Field(
        None,
        description="Age group filter. Allowed: 'kids' or 'adults'. If omitted, no age filter is applied.",
    )

    @field_validator("max_price")
    def validate_price_range(cls, v: Optional[float], info: Any) -> Optional[float]:
        """Ensure max_price >= min_price if both are provided."""
        if v is not None and hasattr(info, "data") and "min_price" in info.data:
            min_price = info.data["min_price"]
            if min_price is not None and v < min_price:
                raise ValueError("max_price must be greater than or equal to min_price")
        return v

    @field_validator("search")
    def clean_string_fields(cls, v: Optional[str]) -> Optional[str]:
        """Clean and normalize string fields."""
        if v is not None:
            return v.strip()
        return v

    @field_validator("age_group")
    def validate_age_group(cls, v: Optional[str]) -> Optional[str]:
        # Treat None or empty string as "no filter"
        if v is None or v == "":
            return None
        vv = str(v).strip().lower()
        # Also treat stripped empty or 'both' as no filter
        if vv == "" or vv == "both":
            return None
        if vv not in {"kids", "adults"}:
            raise ValueError("age_group must be one of: 'kids', 'adults'")
        return vv

    # Enforce strictness for query param schema when instantiated explicitly
    model_config = ConfigDict(extra="forbid", validate_assignment=True)


class PreferredTeachingLocationIn(BaseModel):
    """Preferred teaching location input payload."""

    address: str = Field(..., min_length=1, max_length=512)
    label: str | None = Field(default=None, min_length=1, max_length=64)

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class PreferredPublicSpaceIn(BaseModel):
    """Preferred public space input payload."""

    address: str = Field(..., min_length=1, max_length=512)
    label: str | None = Field(default=None, min_length=1, max_length=64)

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class PreferredTeachingLocationOut(PreferredTeachingLocationIn):
    """Preferred teaching location response payload."""

    model_config = ConfigDict(from_attributes=True)

    @model_serializer
    def serialize(self) -> dict[str, Any]:
        data: dict[str, Any] = {"address": self.address}
        if self.label is not None:
            data["label"] = self.label
        return data


class PreferredPublicSpaceOut(PreferredPublicSpaceIn):
    """Preferred public space response payload."""

    model_config = ConfigDict(from_attributes=True)

    @model_serializer
    def serialize(self) -> dict[str, Any]:
        data: dict[str, Any] = {"address": self.address}
        if self.label is not None:
            data["label"] = self.label
        return data


class ServiceBase(StandardizedModel):
    """
    Base schema for instructor services.

    Attributes:
        service_catalog_id: ID of the service from the catalog
        hourly_rate: Rate per hour in USD
        description: Optional description of the service
        duration_options: Available duration options for this service in minutes
    """

    service_catalog_id: str = Field(..., description="ID of the service from catalog")
    hourly_rate: Money = Field(
        ..., gt=0, le=1000, description="Hourly rate in USD"
    )  # Changed from float
    description: Optional[str] = Field(None, max_length=500)
    requirements: Optional[str] = Field(None, max_length=500)
    age_groups: Optional[List[str]] = Field(
        default=None,
        description="Age groups this service is offered to. Allowed: 'kids', 'adults'. Use both for both.",
    )
    levels_taught: Optional[List[str]] = Field(
        default=None,
        description="Levels taught. Allowed: 'beginner', 'intermediate', 'advanced'",
    )
    equipment_required: Optional[List[str]] = Field(
        default=None,
        description="List of equipment required (strings)",
    )
    location_types: Optional[List[str]] = Field(
        default=None,
        description="Where lessons are offered. Allowed: 'in-person', 'online'",
    )
    duration_options: List[int] = Field(
        default=[60],
        description="Available duration options for this service in minutes",
        min_length=1,
    )

    @field_validator("duration_options")
    def validate_duration_range(cls, v: List[int]) -> List[int]:
        """Ensure each duration is within valid range."""
        if not v:
            raise ValueError("At least one duration option is required")
        for duration in v:
            if not MIN_SESSION_DURATION <= duration <= MAX_SESSION_DURATION:
                raise ValueError(
                    f"Duration must be between {MIN_SESSION_DURATION} and {MAX_SESSION_DURATION} minutes"
                )
        return v

    @field_validator("age_groups")
    def validate_age_groups(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        """Normalize and validate age groups.

        Accepts ['kids'], ['adults'], or ['kids','adults'].
        Collapses duplicates and ignores casing. Raises if unknown values provided.
        """
        if v is None:
            return v
        allowed = {"kids", "adults"}
        normalized: List[str] = []
        for item in v:
            value = str(item).strip().lower()
            if value == "both":
                normalized.extend(["kids", "adults"])
            elif value in allowed:
                normalized.append(value)
            else:
                raise ValueError("age_groups must be one or more of: 'kids', 'adults'")
        # De-duplicate while preserving order
        seen: set[str] = set()
        deduped: List[str] = []
        for item in normalized:
            if item not in seen:
                seen.add(item)
                deduped.append(item)
        return deduped

    @field_validator("levels_taught")
    def validate_levels_taught(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        """Normalize and validate levels_taught.

        Allowed values: beginner, intermediate, advanced.
        """
        if v is None:
            return v
        allowed = {"beginner", "intermediate", "advanced"}
        normalized: List[str] = []
        for item in v:
            value = str(item).strip().lower()
            if value not in allowed:
                raise ValueError(
                    "levels_taught must be any of: 'beginner', 'intermediate', 'advanced'"
                )
            if value not in normalized:
                normalized.append(value)
        return normalized

    @field_validator("location_types")
    def validate_location_types(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        """Normalize and validate location types.

        Allowed values: in-person, online.
        """
        if v is None:
            return v
        allowed = {"in-person", "online"}
        normalized: List[str] = []
        for item in v:
            value = str(item).strip().lower()
            if value not in allowed:
                raise ValueError("location_types must be one or more of: 'in-person', 'online'")
            if value not in normalized:
                normalized.append(value)
        return normalized


class ServiceCreate(StrictRequestModel, ServiceBase):
    """Schema for creating a new service."""

    # Preserve existing response config (use_enum_values/populate_by_name) while forbidding extras
    model_config = ConfigDict(
        **ServiceBase.model_config,
        **StrictRequestModel.model_config,
    )


class ServiceResponse(ServiceBase):
    """
    Schema for service responses.

    Includes the service ID and catalog information.
    """

    id: str
    name: str | None = Field(
        default=None,
        description="Resolved name of the service from the catalog",
    )
    service_catalog_name: str = Field(
        ...,
        max_length=255,
        description="Human-readable name of the catalog service",
    )
    display_order: int | None = Field(
        default=None,
        description="Display order hint from the catalog (nullable)",
    )
    online_capable: bool | None = Field(
        default=None,
        description="Whether the catalog service supports online delivery",
    )
    requires_certification: bool | None = Field(
        default=None,
        description="Whether the catalog service requires certification",
    )
    is_active: bool | None = Field(
        default=None,
        description="Whether this service is currently active for the instructor",
    )

    model_config = ConfigDict(from_attributes=True, extra="forbid", validate_assignment=True)


class UserBasic(StandardizedModel):
    """Basic user information for embedding in responses."""

    first_name: str
    last_name: str
    email: str

    model_config = ConfigDict(from_attributes=True)


class UserBasicPrivacy(StandardizedModel):
    """
    Basic user information with privacy protection.

    Shows only last initial instead of full last name for privacy.
    Used in student-facing endpoints to protect instructor privacy.
    Email is omitted for privacy protection.
    """

    id: str
    first_name: str
    last_initial: str  # Only last initial, not full last_name
    # No email field for privacy protection

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_user(cls, user: Any) -> "UserBasicPrivacy":
        """
        Create UserBasicPrivacy from User model.
        Ensures privacy by only exposing last initial and no email.

        Args:
            user: User model object

        Returns:
            UserBasicPrivacy instance with protected data
        """
        return cls(
            id=user.id,
            first_name=user.first_name,
            last_initial=user.last_name[0] if user.last_name else "",
            # No email for privacy protection
        )


class InstructorProfileBase(StandardizedModel):
    """
    Base schema for instructor profiles.

    Note: Removed fields from old booking system:
        - default_session_duration (moved to service level)
        - buffer_time (will be in booking settings)
        - minimum_advance_hours (will be in booking settings)
    """

    bio: str = Field(
        ...,
        min_length=MIN_BIO_LENGTH,
        max_length=MAX_BIO_LENGTH,
        description="Instructor biography/description",
    )
    years_experience: int = Field(..., ge=0, le=50, description="Years of teaching experience")
    min_advance_booking_hours: int = Field(
        default=2, ge=0, le=168, description="Minimum hours in advance for bookings"
    )
    buffer_time_minutes: int = Field(
        default=0, ge=0, le=60, description="Buffer time between bookings"
    )

    @field_validator("bio")
    def validate_bio(cls, v: str) -> str:
        """Ensure bio is not just whitespace."""
        if not v.strip():
            raise ValueError("Bio cannot be empty")
        return v.strip()


class InstructorProfileCreate(StrictRequestModel, InstructorProfileBase):
    """
    Schema for creating an instructor profile.

    Requires at least one service to be defined.
    """

    services: List[ServiceCreate] = Field(
        ...,
        min_length=1,
        max_length=20,
        description="Services offered by the instructor",
    )

    @field_validator("services")
    def validate_unique_services(cls, v: List[ServiceCreate]) -> List[ServiceCreate]:
        """Ensure no duplicate service catalog IDs."""
        catalog_ids = [service.service_catalog_id for service in v]
        if len(catalog_ids) != len(set(catalog_ids)):
            raise ValueError("Duplicate services are not allowed")
        return v

    # Maintain populate_by_name/use_enum_values from base while enforcing strict requests
    model_config = ConfigDict(
        **InstructorProfileBase.model_config,
        **StrictRequestModel.model_config,
    )


class InstructorProfileUpdate(StrictRequestModel):
    """
    Schema for updating an instructor profile.

    All fields are optional for partial updates.
    """

    bio: Optional[str] = Field(None, min_length=MIN_BIO_LENGTH, max_length=MAX_BIO_LENGTH)
    years_experience: Optional[int] = Field(None, ge=0, le=50)
    services: Optional[List[ServiceCreate]] = Field(None, min_length=0, max_length=20)
    min_advance_booking_hours: Optional[int] = Field(
        None, ge=0, le=168, description="Minimum hours in advance for bookings"
    )
    buffer_time_minutes: Optional[int] = Field(
        None, ge=0, le=60, description="Buffer time between bookings"
    )
    preferred_teaching_locations: Optional[List[PreferredTeachingLocationIn]] = Field(default=None)
    preferred_public_spaces: Optional[List[PreferredPublicSpaceIn]] = Field(default=None)

    @field_validator("preferred_teaching_locations")
    def validate_preferred_teaching_locations(
        cls, v: Optional[List[PreferredTeachingLocationIn]]
    ) -> Optional[List[PreferredTeachingLocationIn]]:
        if v is not None and len(v) > 2:
            raise ValueError("preferred_teaching_locations must contain at most two entries")
        return v

    @field_validator("preferred_public_spaces")
    def validate_preferred_public_spaces(
        cls, v: Optional[List[PreferredPublicSpaceIn]]
    ) -> Optional[List[PreferredPublicSpaceIn]]:
        if v is not None and len(v) > 2:
            raise ValueError("preferred_public_spaces must contain at most two entries")
        return v

    @field_validator("services")
    def validate_unique_services(
        cls, v: Optional[List[ServiceCreate]]
    ) -> Optional[List[ServiceCreate]]:
        """Ensure no duplicate service catalog IDs if provided."""
        if v is not None:
            catalog_ids = [service.service_catalog_id for service in v]
            if len(catalog_ids) != len(set(catalog_ids)):
                raise ValueError("Duplicate services are not allowed")
        return v


class InstructorProfileResponse(InstructorProfileBase):
    """
    Schema for instructor profile responses with privacy protection.

    Includes all profile data plus relationships and metadata.
    Student-facing endpoints will show only instructor last initial.
    """

    id: str
    user_id: str
    created_at: datetime
    updated_at: Optional[datetime] = None
    user: UserBasicPrivacy  # Changed from UserBasic to protect privacy
    services: List[ServiceResponse]
    is_favorited: Optional[bool] = Field(
        None, description="Whether the current user has favorited this instructor"
    )
    favorited_count: int = Field(0, description="Number of students who favorited this instructor")
    # Onboarding status fields
    skills_configured: bool = Field(
        default=False, description="Whether skills/pricing were configured at least once"
    )
    identity_verified_at: Optional[datetime] = Field(default=None)
    identity_verification_session_id: Optional[str] = Field(default=None)
    background_check_object_key: Optional[str] = Field(default=None)
    background_check_uploaded_at: Optional[datetime] = Field(default=None)
    onboarding_completed_at: Optional[datetime] = Field(default=None)
    is_live: bool = Field(default=False)
    is_founding_instructor: bool = Field(
        default=False, description="Whether the instructor is a founding instructor"
    )
    preferred_teaching_locations: List[PreferredTeachingLocationOut] = Field(default_factory=list)
    preferred_public_spaces: List[PreferredPublicSpaceOut] = Field(default_factory=list)
    service_area_neighborhoods: List[ServiceAreaNeighborhoodOut] = Field(default_factory=list)
    service_area_boroughs: List[str] = Field(default_factory=list)
    service_area_summary: Optional[str] = Field(default=None)

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_orm(cls, instructor_profile: Any) -> "InstructorProfileResponse":
        """
        Create InstructorProfileResponse from ORM model with privacy protection.

        Args:
            instructor_profile: InstructorProfile ORM object with user relationship

        Returns:
            InstructorProfileResponse with privacy-protected user data
        """
        preferred_places: list[Any] = []
        if hasattr(instructor_profile, "user") and instructor_profile.user is not None:
            preferred_places = getattr(instructor_profile.user, "preferred_places", []) or []

        teaching_locations: List[PreferredTeachingLocationOut] = []
        public_spaces: List[PreferredPublicSpaceOut] = []

        if preferred_places:
            teaching_sorted = sorted(
                [p for p in preferred_places if getattr(p, "kind", None) == "teaching_location"],
                key=lambda place: getattr(place, "position", 0),
            )
            public_sorted = sorted(
                [p for p in preferred_places if getattr(p, "kind", None) == "public_space"],
                key=lambda place: getattr(place, "position", 0),
            )

            for place in teaching_sorted:
                teaching_locations.append(
                    PreferredTeachingLocationOut(
                        address=getattr(place, "address", ""),
                        label=getattr(place, "label", None),
                    )
                )

            for place in public_sorted:
                public_spaces.append(
                    PreferredPublicSpaceOut(
                        address=getattr(place, "address", ""),
                    )
                )

        # Normalize instructor services with catalog names eagerly resolved
        services_source: List[Any] = []
        if hasattr(instructor_profile, "instructor_services"):
            services_source = list(getattr(instructor_profile, "instructor_services", []) or [])
        elif hasattr(instructor_profile, "services"):
            services_source = list(getattr(instructor_profile, "services", []) or [])

        services_data: List[ServiceResponse] = []
        for service in sorted(services_source, key=lambda s: getattr(s, "service_catalog_id", "")):
            catalog_entry = getattr(service, "catalog_entry", None)
            catalog_name = (
                getattr(catalog_entry, "name", None) if catalog_entry is not None else None
            )
            service_payload = ServiceResponse(
                id=getattr(service, "id"),
                service_catalog_id=getattr(service, "service_catalog_id"),
                service_catalog_name=catalog_name or "Unknown Service",
                hourly_rate=getattr(service, "hourly_rate"),
                description=getattr(service, "description", None),
                requirements=getattr(service, "requirements", None),
                age_groups=getattr(service, "age_groups", None),
                levels_taught=getattr(service, "levels_taught", None),
                equipment_required=getattr(service, "equipment_required", None),
                location_types=getattr(service, "location_types", None),
                duration_options=getattr(service, "duration_options", None) or [60],
            )
            services_data.append(service_payload)

        boroughs: set[str] = set()
        neighborhoods_payload: List[ServiceAreaNeighborhoodOut] = []

        neighborhoods_source = getattr(instructor_profile, "service_area_neighborhoods", None)

        if neighborhoods_source:
            for entry in neighborhoods_source:
                if isinstance(entry, dict):
                    neighborhood_id = entry.get("neighborhood_id") or entry.get("id")
                    ntacode = entry.get("ntacode") or entry.get("region_code")
                    name = entry.get("name") or entry.get("region_name")
                    borough = entry.get("borough") or entry.get("parent_region")
                else:
                    neighborhood_id = getattr(entry, "neighborhood_id", None) or getattr(
                        entry, "id", None
                    )
                    ntacode = getattr(entry, "ntacode", None) or getattr(entry, "region_code", None)
                    name = getattr(entry, "name", None) or getattr(entry, "region_name", None)
                    borough = getattr(entry, "borough", None) or getattr(
                        entry, "parent_region", None
                    )

                neighborhoods_payload.append(
                    ServiceAreaNeighborhoodOut(
                        neighborhood_id=str(neighborhood_id) if neighborhood_id else "",
                        ntacode=ntacode,
                        name=name,
                        borough=borough,
                    )
                )
                if borough:
                    boroughs.add(borough)
        else:
            user_service_areas: Sequence[Any] = []
            if hasattr(instructor_profile, "user") and instructor_profile.user is not None:
                user_service_areas = getattr(instructor_profile.user, "service_areas", []) or []

            for area in user_service_areas:
                neighborhood = getattr(area, "neighborhood", None)
                neighborhood_id = getattr(area, "neighborhood_id", None)
                if neighborhood is None:
                    continue

                borough = getattr(neighborhood, "parent_region", None) or getattr(
                    neighborhood, "borough", None
                )
                ntacode = getattr(neighborhood, "region_code", None) or getattr(
                    neighborhood, "ntacode", None
                )
                name = getattr(neighborhood, "region_name", None) or getattr(
                    neighborhood, "name", None
                )
                neighborhoods_payload.append(
                    ServiceAreaNeighborhoodOut(
                        neighborhood_id=str(getattr(neighborhood, "id", neighborhood_id or "")),
                        ntacode=ntacode,
                        name=name,
                        borough=borough,
                    )
                )
                if borough:
                    boroughs.add(borough)

        sorted_boroughs = sorted(boroughs)
        if sorted_boroughs:
            if len(sorted_boroughs) <= 2:
                service_area_summary = ", ".join(sorted_boroughs)
            else:
                service_area_summary = f"{sorted_boroughs[0]} + {len(sorted_boroughs) - 1} more"
        else:
            service_area_summary = None

        neighborhoods_output = [entry.model_dump(mode="python") for entry in neighborhoods_payload]

        return cls(
            id=instructor_profile.id,
            user_id=instructor_profile.user_id,
            created_at=instructor_profile.created_at,
            updated_at=instructor_profile.updated_at,
            bio=instructor_profile.bio,
            years_experience=instructor_profile.years_experience,
            min_advance_booking_hours=instructor_profile.min_advance_booking_hours,
            buffer_time_minutes=instructor_profile.buffer_time_minutes,
            user=UserBasicPrivacy.from_user(instructor_profile.user),
            services=services_data,
            is_favorited=getattr(instructor_profile, "is_favorited", None),
            favorited_count=getattr(instructor_profile, "favorited_count", 0),
            skills_configured=getattr(instructor_profile, "skills_configured", False),
            identity_verified_at=getattr(instructor_profile, "identity_verified_at", None),
            background_check_uploaded_at=getattr(
                instructor_profile, "background_check_uploaded_at", None
            ),
            onboarding_completed_at=getattr(instructor_profile, "onboarding_completed_at", None),
            is_live=getattr(instructor_profile, "is_live", False),
            is_founding_instructor=getattr(instructor_profile, "is_founding_instructor", False),
            preferred_teaching_locations=teaching_locations,
            preferred_public_spaces=public_spaces,
            service_area_neighborhoods=neighborhoods_output,
            service_area_boroughs=sorted_boroughs,
            service_area_summary=service_area_summary,
        )

    @field_validator("services")
    def sort_services(cls, v: List[ServiceResponse]) -> List[ServiceResponse]:
        """Sort services by catalog ID for consistent display."""
        return sorted(v, key=lambda s: s.service_catalog_id)

File: frontend/types/instructor.ts
// frontend/types/instructor.ts

/**
 * Instructor Type Definitions
 *
 * This module contains all TypeScript interfaces and types related to
 * instructors, their profiles, services, and search results.
 *
 * @module instructor
 */

/**
 * Basic instructor information
 *
 * Used in listings, search results, and anywhere a simplified
 * instructor representation is needed.
 *
 * @interface InstructorBasic
 */
export interface ServiceAreaNeighborhood {
  neighborhood_id: string;
  ntacode?: string | null;
  name?: string | null;
  borough?: string | { name?: string | null; label?: string | null } | null;
}

export interface InstructorBasic {
  /** Instructor profile ID (ULID string) */
  id: string;

  /** Associated user ID (ULID string) */
  user_id: string;

  /** Instructor's first name */
  first_name: string;

  /** Instructor's last initial for privacy */
  last_initial: string;

  /** Brief bio/description */
  bio: string;

  /** Founding instructor flag */
  is_founding_instructor?: boolean;

  /** Years of teaching experience */
  years_experience: number;

  /** Ordered list of borough labels derived from neighborhoods */
  service_area_boroughs?: string[];

  /** Detailed neighborhood metadata for service areas */
  service_area_neighborhoods?: ServiceAreaNeighborhood[];

  /** Human-readable summary provided by backend */
  service_area_summary?: string | null;

  /** Average rating (optional, for future use) */
  average_rating?: number;

  /** Total number of reviews (optional, for future use) */
  total_reviews?: number;

  /** Profile image URL (optional, for future use) */
  profile_image_url?: string;
}

/**
 * Complete instructor profile with user information
 *
 * Used in instructor dashboard, detailed profile views, and
 * anywhere full instructor information is needed.
 *
 * @interface InstructorProfile
 * @extends InstructorBasic
 */
export interface InstructorProfile {
  /** Instructor profile ID (ULID string) */
  id: string;

  /** Associated user ID (ULID string) */
  user_id: string;

  /** Detailed bio/description */
  bio: string;

  /** Ordered list of borough labels derived from neighborhoods */
  service_area_boroughs?: string[];

  /** Detailed neighborhood metadata for service areas */
  service_area_neighborhoods?: ServiceAreaNeighborhood[];

  /** Human-readable summary provided by backend */
  service_area_summary?: string | null;

  /** Preferred teaching locations configured by instructor */
  preferred_teaching_locations?: Array<{
    address: string;
    label?: string;
  }>;

  /** Preferred public spaces configured by instructor */
  preferred_public_spaces?: Array<{
    address: string;
    label?: string;
  }>;

  /** Years of teaching experience */
  years_experience: number;

  /** Associated user information */
  user: {
    /** User's first name */
    first_name: string;
    /** User's last initial for privacy */
    last_initial: string;
    /** Whether this user has uploaded a profile picture */
    has_profile_picture?: boolean;
    /** Current profile picture version for cache busting */
    profile_picture_version?: number;
    // Email removed for privacy
  };

  /** Top-level profile picture indicator (mirrors user.has_profile_picture) */
  has_profile_picture?: boolean;

  /** Top-level profile picture version (mirrors user.profile_picture_version) */
  profile_picture_version?: number;

  /** Services offered by this instructor */
  services: InstructorService[];

  /** Whether the current user has favorited this instructor */
  is_favorited?: boolean;

  /** Total number of students who favorited this instructor */
  favorited_count: number;

  /** Verification status (for future use) */
  is_verified?: boolean;

  /** Founding instructor flag */
  is_founding_instructor?: boolean;

  /** Background check status (for future use) */
  background_check_completed?: boolean;
  /** Background check Review status literal */
  bgc_status?: string;
  /** Timestamp for completed background check */
  bgc_completed_at?: string | null;

  /** Date when instructor joined */
  created_at?: string;

  // Onboarding status fields
  skills_configured?: boolean;
  identity_verified_at?: string | null;
  identity_verification_session_id?: string | null;
  background_check_object_key?: string | null;
  background_check_uploaded_at?: string | null;
  onboarding_completed_at?: string | null;
  is_live?: boolean;
}

/**
 * Service offered by an instructor
 *
 * Represents a specific skill/service that an instructor teaches,
 * including pricing and optional details.
 *
 * @interface InstructorService
 */
export interface InstructorService {
  /** Service ID (ULID string) */
  id: string;

  /** Service catalog ID (ULID string) */
  service_catalog_id?: string;

  /** Human readable service name provided by the API */
  service_catalog_name?: string;

  /** Skill name (e.g., "Piano", "Yoga", "Spanish") */
  skill?: string;

  /** Hourly rate in USD */
  hourly_rate: number;

  /** Optional service description */
  description: string | null;

  /** Available duration options in minutes */
  duration_options?: number[];

  /** Modalities / delivery methods supported (e.g., 'in-person', 'online') */
  location_types?: string[];

  /** Whether this service is currently active */
  is_active?: boolean;

  /** Age groups this service is offered to (e.g., 'kids', 'adults') */
  age_groups?: string[];

  /** Skill levels taught (e.g., 'beginner', 'intermediate', 'advanced') */
  levels_taught?: string[];

  /** Instructor ID (ULID string, when not nested) */
  instructor_id?: string;
}

/**
 * Instructor search/browse result
 *
 * Used in the browse instructors page and search results,
 * combining instructor info with relevant services.
 *
 * @interface InstructorSearchResult
 */
export interface InstructorSearchResult {
  /** Basic instructor information */
  instructor: InstructorBasic;

  /** Services offered (filtered by search if applicable) */
  services: InstructorService[];

  /** Next available slot (ISO datetime) */
  next_available?: string;

  /** Distance from user (for future location-based search) */
  distance_miles?: number;

  /** Match score (for future search ranking) */
  relevance_score?: number;
}

// InstructorListResponse removed - use generated type from @/features/shared/api/types

/**
 * Request parameters for searching instructors
 *
 * @interface InstructorSearchParams
 */
export interface InstructorSearchParams {
  /** Search query (matches name, bio, skills) */
  query?: string;

  /** Filter by specific skill */
  skill?: string;

  /** Filter by area of service */
  area?: string;

  /** Minimum hourly rate */
  min_rate?: number;

  /** Maximum hourly rate */
  max_rate?: number;

  /** Minimum years of experience */
  min_experience?: number;

  /** Only show instructors with availability */
  available_only?: boolean;

  /** Sort order */
  sort_by?: 'rating' | 'price_low' | 'price_high' | 'experience' | 'distance';

  /** Page number for pagination */
  page?: number;

  /** Results per page */
  per_page?: number;
}

/**
 * Instructor statistics (for dashboard)
 *
 * @interface InstructorStats
 */
export interface InstructorStats {
  /** Total number of bookings */
  total_bookings: number;

  /** Completed bookings */
  completed_bookings: number;

  /** Cancelled bookings */
  cancelled_bookings: number;

  /** No-show bookings */
  no_show_bookings: number;

  /** Total earnings */
  total_earnings: number;

  /** Average rating */
  average_rating: number;

  /** Total reviews */
  total_reviews: number;

  /** This month's earnings */
  monthly_earnings: number;

  /** This week's bookings */
  weekly_bookings: number;

  /** Upcoming bookings count */
  upcoming_bookings: number;
}

/**
 * Favorited instructor from the favorites API
 *
 * @interface FavoritedInstructor
 */
export interface FavoritedInstructor {
  /** Instructor user ID (ULID string) */
  id: string;

  /** Instructor email */
  email: string;

  /** Instructor first name */
  first_name: string;

  /** Instructor last name */
  last_name: string;

  /** Whether the instructor is active */
  is_active: boolean;

  /** Instructor profile details */
  profile?: InstructorProfile;

  /** When this instructor was favorited */
  favorited_at?: string;

  /** Whether instructor has profile picture (optional API field) */
  has_profile_picture?: boolean;

  /** Profile picture version (optional API field) */
  profile_picture_version?: number;
}

// FavoritesListResponse removed - use generated type from @/features/shared/api/types

/**
 * Type guard to check if a user has an instructor profile
 *
 * @param user - User object to check
 * @returns boolean indicating if user is an instructor
 */
export function isInstructor(user: unknown): user is { instructor_profile: InstructorProfile } {
  return Boolean(user && typeof user === 'object' && 'instructor_profile' in user);
}

/**
 * Helper to format instructor display name
 *
 * @param instructor - Instructor data
 * @returns Formatted display name
 */
export function getInstructorDisplayName(instructor: unknown): string {
  const inst = instructor as Record<string, unknown>;
  // For InstructorProfile with nested user object
  if (inst?.['user'] && typeof inst['user'] === 'object') {
    const user = inst['user'] as Record<string, unknown>;
    if (user?.['first_name'] && user?.['last_initial']) {
      const firstName = String(user['first_name']) || '';
      const lastInitial = String(user['last_initial']) || '';
      return lastInitial ? `${firstName} ${lastInitial}.` : firstName;
    }
  }
  // For InstructorBasic with direct fields
  if (inst?.['first_name'] && inst?.['last_initial']) {
    const firstName = String(inst['first_name']) || '';
    const lastInitial = String(inst['last_initial']) || '';
    return lastInitial ? `${firstName} ${lastInitial}.` : firstName;
  }
  // Fallback
  if (inst?.['user_id']) {
    return 'Instructor #' + String(inst['user_id']);
  }
  return 'Instructor';
}

/**
 * Helper to get primary service (lowest priced or first)
 *
 * @param services - Array of services
 * @returns Primary service or null
 */
export function getPrimaryService(services: InstructorService[]): InstructorService | null {
  if (!services || services.length === 0) return null;

  return services.reduce((primary, service) =>
    service.hourly_rate < primary.hourly_rate ? service : primary
  );
}
